docs/design/persistent-storage.md

```diff
@@ -0,0 +1,220 @@
+# Persistent Storage
+
+This document proposes a model for managing persistent, cluster-scoped storage for applications requiring long lived data.
+
+### tl;dr
+
+`PersistentStorage` objects contain a VolumeSource with a lifecycle greater than the pods it is used by.  Many types of VolumeSources can be implemented as volume plugins, with each implementing an interface for use within a controller.  A pod with persistence requirements (specifically pod.spec.volumes.volume.persistenceRequirements) is bound to a matching VolumeSource from a pool of available storage.  PersistentStorage objects give that VolumeSource longevity.
+
+A `PersistentStorageController` manages the creation, attachment, and reclamation of PersistentStorage objects for one VolumeSource.  Implementations must be pluggable to support many types of controllers for many types of VolumeSources. The controller can advise the scheduler on host selection, keeping volume rules contained within the provider.  All cloud provider credentials are isolated to the storage controller's host.
+
+`PersistentStorageBinder` watches for pods with persistence requirements and matches the request with what is available in the pool.
+
+`PersistenceRequirements` expresses the capabilities and performance attributes of the underlying storage.  Allow discovery
+of storage by its capabilities and use those capabilities as pod storage requirements.
+
+Kubernetes makes no guarantees at runtime that the underlying storage exists or is available.  High availability is left to the storage provider.
+
+## Goals
+
+* Allow administrators to describe available storage
+* Allow pod authors to discover and request persistence for their volumes
+* Enforce security through access control lists and securing storage to the same namespace as the pod volume
+* Enforce quotas through admission control
+* Ensure developers can rely on storage being available without being closely bound to a particular disk, server, network, or storage device.
+
+
+#### Allow admins to describe and create available storage
+
+Admins can describe storage with `PersistentStorageControllers` or create storage directly by using the `PersistentStorage` API.  Storage created manually still requires a PersistentStorageController and accompanying interface to allow management by the cluster.
+
+
+```
+	
+	PersistentStorage
+		TypeMeta
+		ObjectMeta
+			Name - internal name
+			Labels - labeled like Volume.PersistenceRequirements.Selector
+		Spec
+			VolumeSource
+				HostPath
+				EmptyDir
+				GitRepo	
+				GCEPersistentDisk { modes: ReadWriteOnce | ReadOnlyMany }
+				AWSElasticBlockStore { modes: ReadWriteOnce }
+				NFSMount { modes: ReadWriteOnce | ReadOnlyMany | ReadWriteMany }
+				*** many volume plugins
```

Comment:
I'd like to factor out reasonably common properties that are necessary for binding. Everything else shouldn't be part of the API -- they should be plugins. I don't want anything in the API that names a specific cloud provider. There just needs to be enough information so that a cloudprovider implementation can attach the provisioned device to the node, and then so that the device can be mounted (so device uuid or something in status).